Hackathon (Core) - Telemetry and analytics

jgambarios opened the pull request:
Focusing on adding new event types and sources, and enhancing the CLI tool with analytics tracking capabilities. The most important changes include adding new event sources and types, creating a new analytics API, and modifying the CLI tool to support event recording.

### Event Source and Type Additions:
* [`dotCMS/src/main/java/com/dotcms/analytics/track/collectors/EventSource.java`](diffhunk://#diff-9d75ec20f52e88f8038bd099b714db99edd94d2ccc670c74d5167bed5322d198L12-R13): Added `DOT_CLI` as a new event source.
* [`dotCMS/src/main/java/com/dotcms/analytics/track/collectors/EventType.java`](diffhunk://#diff-7eb71fe1109ae1749f4896e8ce1e62d82a1a2651ba64d4e48db1748c0a85864fR12-L13): Added `CMD_EXECUTED` as a new event type.

### Analytics API Implementation:
* [`tools/dotcms-cli/api-data-model/src/main/java/com/dotcms/api/AnalyticsAPI.java`](diffhunk://#diff-dbf9f08f558d4307b10a0bddaaaed4519a0ccaf0d1c910505caa6b555b969d52R1-R35): Created a new `AnalyticsAPI` interface to handle custom user events.
* [`tools/dotcms-cli/api-data-model/src/main/java/com/dotcms/model/analytics/AbstractDotCliEvent.java`](diffhunk://#diff-e6553f0644fa6666f14d86d2e6205287a7a4d924f1b2b70cfd6f5a58a4d61c26R1-R17): Defined a new `AbstractDotCliEvent` interface for CLI event data.

### CLI Tool Enhancements:
* [`tools/dotcms-cli/cli/src/main/java/com/dotcms/cli/command/CustomConfigurationUtil.java`](diffhunk://#diff-835a02999ed76e445ab45df1908bd97edc5c82bed5f42bad7bb8f2422a77d4e0R3): Injected `RestClientFactory` to support the new analytics API. [[1]](diffhunk://#diff-835a02999ed76e445ab45df1908bd97edc5c82bed5f42bad7bb8f2422a77d4e0R3) [[2]](diffhunk://#diff-835a02999ed76e445ab45df1908bd97edc5c82bed5f42bad7bb8f2422a77d4e0R15) [[3]](diffhunk://#diff-835a02999ed76e445ab45df1908bd97edc5c82bed5f42bad7bb8f2422a77d4e0R26-R28) [[4]](diffhunk://#diff-835a02999ed76e445ab45df1908bd97edc5c82bed5f42bad7bb8f2422a77d4e0L37-R43)
* [`tools/dotcms-cli/cli/src/main/java/com/dotcms/cli/common/DotExecutionStrategy.java`](diffhunk://#diff-690c10957cfe8077e8f2cb8362b675fbe965adf1019114280b13c33df7d906ebR3-R24): Modified the execution strategy to record events asynchronously using the new analytics API. [[1]](diffhunk://#diff-690c10957cfe8077e8f2cb8362b675fbe965adf1019114280b13c33df7d906ebR3-R24) [[2]](diffhunk://#diff-690c10957cfe8077e8f2cb8362b675fbe965adf1019114280b13c33df7d906ebR51) [[3]](diffhunk://#diff-690c10957cfe8077e8f2cb8362b675fbe965adf1019114280b13c33df7d906ebR65-R80) [[4]](diffhunk://#diff-690c10957cfe8077e8f2cb8362b675fbe965adf1019114280b13c33df7d906ebL87-R178) [[5]](diffhunk://#diff-690c10957cfe8077e8f2cb8362b675fbe965adf1019114280b13c33df7d906ebR268-R300)

### Additional changes:
* ~Tracking analytics on Future Time Machine~

This PR fixes: #31312

wezell committed:
bug(apps) fixing link to dotAI portlet (#29556)

ref: #29555
Changing link from `/c/dotAI` to `/c/dotai`

valentinogiardino committed:
#31312 add API for analytics dotcli custom event

jgambarios committed:
#31312 Integrating analytics to the CLI

valentinogiardino committed:
#31312 fix remove event_source

nollymar committed:
#31312 Moving Content Analytics event registration for rest events to an utility class

nollymar committed:
#31312 Registering analytics event for FTM requests

jgambarios committed:
feat(core): enhance CLI command execution with parallel processing and event recording

jcastro-dotcms committed:
Fixes #31312 : Adding some of the new AI Telemetry Metrics

nollymar committed:
#31312 Improving payload of content analytics event

jdotcms commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/analytics/content/util/ContentAnalyticsUtil.java:
```diff
@@ -0,0 +1,71 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class ContentAnalyticsUtil {
```
Doc me and probably helper will be a more accurate name

jdotcms commented on dotCMS/src/main/java/com/dotcms/telemetry/collectors/MetricStatsCollector.java:
```diff
@@ -223,6 +226,10 @@ private MetricStatsCollector() {
         metricStatsCollectors.add(new CountOfTimeFieldsMetricType());
         metricStatsCollectors.add(new CountOfWYSIWYGFieldsMetricType());
 
+        metricStatsCollectors.add(new TotalSitesUsingDotaiMetricType());
```
Rename to DotAI

jdotcms commented on dotCMS/src/main/java/com/dotcms/telemetry/collectors/ai/TotalEmbeddingsIndexesMetricType.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class TotalEmbeddingsIndexesMetricType implements MetricType {
```
Doc

jdotcms commented on dotCMS/src/main/java/com/dotcms/telemetry/collectors/ai/TotalSitesUsingDotaiMetricType.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Optional;
+import java.util.Set;
+
+public class TotalSitesUsingDotaiMetricType implements MetricType {
```
Doc me

jdotcms commented on dotCMS/src/main/java/com/dotcms/telemetry/collectors/ai/TotalSitesUsingDotaiMetricType.java:
```diff
@@ -0,0 +1,67 @@
+            return Optional.of(new MetricValue(this.getMetric(), sitesWithConfigurations.size()));
+        } catch (final DotSecurityException e) {
+            throw new DotDataException(e);
+        }
```
finally here to destroy the secrets

jdotcms commented on dotCMS/src/main/java/com/dotcms/telemetry/collectors/ai/TotalSitesWithAutoIndexContentConfigMetricType.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.Optional;
+import java.util.Set;
+
+public class TotalSitesWithAutoIndexContentConfigMetricType implements MetricType {
```
Docme

jdotcms commented on dotCMS/src/main/java/com/dotcms/telemetry/collectors/ai/TotalSitesWithAutoIndexContentConfigMetricType.java:
```diff
@@ -0,0 +1,83 @@
+            return Optional.of(new MetricValue(this.getMetric(), counter));
+        } catch (final DotSecurityException e) {
+            throw new DotDataException(e);
+        }
```
finally delete secrets

jcastro-dotcms committed:
Adding some more code changes

jcastro-dotcms committed:
Merge remote-tracking branch 'origin/issue-31312-Hackathon-Every-feature-we-build-has-telemetry' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

nollymar committed:
Merge remote-tracking branch 'origin' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

nollymar committed:
#31312 Removing unnecessary fields from the analytics event payload

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

jgambarios committed:
#31312 Refactor analytics handling and improve event recording clarity

Replaced direct usage of AnalyticsAPI with a dedicated AnalyticsService to decouple responsibilities and streamline event recording. Improved logging for better debugging and enhanced integration of analytics commands with argument passing for more accurate tracking. Removed unused interfaces and adapted existing code for the new service structure.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

jgambarios committed:
#31312 Refactor analytics handling and improve event recording clarity

Replaced direct usage of AnalyticsAPI with a dedicated AnalyticsService to decouple responsibilities and streamline event recording. Improved logging for better debugging and enhanced integration of analytics commands with argument passing for more accurate tracking. Removed unused interfaces and adapted existing code for the new service structure.

jgambarios committed:
#31312 Add support for InstanceCommand in execution strategy

Updated `DotExecutionStrategy` to include `InstanceCommand` in allowed commands. Made `InstanceCommand.NAME` public for broader accessibility and replaced a DI field with dynamic instantiation of `AnalyticsService` for consistency.

jgambarios committed:
#31312 Improve analytics service integration in CLI commands

Refactor to use dependency injection for `AnalyticsService` across CLI classes, improving flexibility and consistency. Added logging to warn when the analytics service is unavailable, ensuring better error visibility during command execution. Adjusted `AnalyticsServiceImpl` to use constructor-based injection for better design.

jgambarios committed:
#31312 Enable analytics tracking configuration in CLI

Added a configuration property to toggle analytics tracking. Refactored code to dynamically retrieve the AnalyticsService instance when tracking is enabled, removing its injection and related constructor dependencies. This ensures commands are logged only if analytics are explicitly enabled.

jgambarios committed:
#31312 Avoid recording analytics events during test execution.

This change adds a check for the presence of a test profile before attempting to record analytics events. It ensures that the analytics service is not invoked when running tests, preventing unnecessary overhead and potential interference with test results.

jgambarios committed:
#31312 Avoid recording analytics events during test execution.

This change adds a check for the presence of a test profile before attempting to record analytics events. It ensures that the analytics service is not invoked when running tests, preventing unnecessary overhead and potential interference with test results.

jgambarios committed:
Test commit

jgambarios committed:
#31312 Add ManagedExecutor for async task handling in CLI commands

This update integrates `ManagedExecutor` to manage asynchronous tasks within `DotExecutionStrategy` and related areas. It improves task execution by ensuring thread safety and leveraging the MicroProfile Concurrency API. Additionally, analytics event recording is updated to respect configuration settings and ignore specific commands.

jgambarios committed:
Merge remote-tracking branch 'origin/master' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

jgambarios committed:
#31312 Add ManagedExecutor for async task handling in CLI commands

This update integrates `ManagedExecutor` to manage asynchronous tasks within `DotExecutionStrategy` and related areas. It improves task execution by ensuring thread safety and leveraging the MicroProfile Concurrency API. Additionally, analytics event recording is updated to respect configuration settings and ignore specific commands.

jgambarios committed:
#31312 Ignoring broken test

jgambarios committed:
#31312 Refactor command execution handling and improve logging.

Refactored the command execution flow by introducing better thread-local management, handling CompletableFuture interruptions more safely, and ensuring proper ThreadLocal cleanup. Enhanced logging to include both the command and its arguments for better debugging and traceability. Also, updated test cases for simplified file path handling and re-enabled a previously disabled test.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

# Conflicts:
#	dotCMS/src/main/java/com/dotcms/analytics/track/collectors/EventType.java
#	dotCMS/src/main/java/com/dotcms/telemetry/MetricFeature.java
#	dotCMS/src/main/java/com/dotcms/telemetry/collectors/MetricStatsCollector.java

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

jgambarios committed:
#31312 Add analytics service tests and improve exception handling

This commit introduces unit and integration tests for the `AnalyticsService`, ensuring proper recording of events and handling of various scenarios. It also enhances `DotExceptionHandler` to gracefully handle `InterruptedException` and adds a minor improvement to the RemoteURLIT test.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-31312-Hackathon-Every-feature-we-build-has-telemetry

nollymar committed:
#31312 Removing analytics logic to track FTM events in BE

nollymar committed:
#31312 Removing analytics logic to track FTM events in BE

nollymar committed:
#31312 Removing unused properties and imports

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/cli/common/DotExecutionStrategy.java:
```diff
@@ -187,6 +303,79 @@ int internalExecute(final CommandsChain commandsChain,
+        final String parentCommand = commandsChain.firstSubcommand()
+                .map(p -> p.commandSpec().name()).orElse("UNKNOWN");
+
+        return !ConfigCommand.NAME.equals(parentCommand)
```
it would be better if we had a an annotation to mark commands as trackable

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/cli/common/DotExecutionStrategy.java:
```diff
@@ -187,6 +303,79 @@ int internalExecute(final CommandsChain commandsChain,
+     * @return true if analytics tracking is enabled; false otherwise
+     */
+    private Boolean isAnalyticsEnabled() {
+
```
Make return type a primitive

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/cli/common/DotExecutionStrategy.java:
```diff
@@ -187,6 +303,79 @@ int internalExecute(final CommandsChain commandsChain,
+     * @return true if the command in the chain should be tracked, false otherwise
+     */
+    private boolean trackCommand(CommandsChain commandsChain) {
+
```
Perhaps rename to shouldTrackCommand or  isTrackableCommand?

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/analytics/AnalyticsServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    @ActivateRequestContext
+    public void recordCommand(final String command, final List<String> arguments)
```
Shouldn't this be asynchronous?  Perhaps we should wrap this within a Future Call so it doesn't impact execution

jgambarios replied:
We call the `AnalyticsServiceImpl` using futures from the `DotExecutionStrategy`.

nollymar committed:
#31312 Removing unused code
